EIPS/eip-wrapped_deposit.md

```diff
@@ -0,0 +1,137 @@
+---
+eip: <to be assigned>
+title: Wrapped Deposits
+description: A singleton contract for managing asset deposits.
+author: Justice Hudson (@jchancehud)
+discussions-to: https://ethereum-magicians.org/t/wrapped-deposit-contract-eip/7740
+status: Draft
+type: Informational
+created: 2021-12-11
+---
+
+## Abstract
+The wrapped deposit contract handles deposits of assets (Ether, ERC20, ERC721) on behalf of a user. A user must only approve a spend limit once and then an asset may be deposited to any number of different applications that support deposits from the contract.
+
+## Motivation
+The current user flow for depositing assets in dapps is unnecessarily expensive and insecure. To deposit an ERC20 asset a user must either:
+
+  - send an approve transaction for the exact amount being sent, before making a deposit; for every deposit.
+  - send an approve transaction for an infinite spend amount before making deposits.
+
+The first option is inconvenient, and expensive. The second option is insecure. Further, explaining approvals to new or non-technical users is confusing. This has to be done in _every_ dapp that supports ERC20 deposits.
```

Comment:
```suggestion
The first option is inconvenient, and expensive. The second option is insecure. Further, explaining approvals to new or non-technical users is confusing. This has to be done in _every_ dapp that supports ERC-20 deposits.
```

Reply:
In addition, wouldn't the singleton face the same two issues? Either the singleton's user gives it an infinite approve, or faces approving each tx separately, no?

Reply:
Sorta, the issue isn't having to do either of these, it's having to do either for _every_ contract I want to interact with. If there's a single contract that manages my deposits for many applications I'm more comfortable setting an infinite spend approval there, especially if I can easily digest the source (and it's not upgrade-able).

Reply:
I'm not sure I see how you get around the insecurity - if there's an infinite approve on your proposed contract, any layer that is interfaced with can still compromise the user's funds via the contract, no?

In addition, each asset would still need an infinite approve, so this doesn't reduce the amount of approvals to one.

Reply:
> any layer that is interfaced with can still compromise the user's funds via the contract

No, the contract is implemented only allowing funds to be deposited on behalf of `msg.sender`. This value changes with any external contract call. So, say I want to deposit to a compromised contract X. Contract X tries to re-enter the deposit contract and spend more of my tokens. Contract X just ends up spending it's own tokens.

The deposit contract should be safe even if the deposit receiver is malicious, this language should probably be in the EIP.

> each asset would still need an infinite approve, so this doesn't reduce the amount of approvals to one

Correct, one approval per asset.